### PR TITLE
[staking] process delegations withdraws at epoch boundaries to include rewards

### PR DESCRIPTION
**Pull request by emmazzz**
This PR changes our staking logic so that 
- Delegation withdraws will also withdraw the rewards from the current epoch. This way our rewards distribution rule is more clear and makes more sense. 
- Same logic applies to switching delegation from one validator to another.
- Some refactoring is also done to make this change easier.

### CODE REVIEW

**Review by lxfind — COMMENTED** [review]
Please add more doc comments describing what each function does semantically. I found it a bit difficult to follow.

File: crates/sui-framework/sources/governance/validator_set.move [hunk 1]
```diff
@@ -184,39 +194,85 @@ module sui::validator_set {
 
     public(friend) fun request_switch_delegation(
         self: &mut ValidatorSet,
-        delegation: Delegation,
+        delegation: &mut Delegation,
         staked_sui: &mut StakedSui,
         new_validator_address: address,
+        switch_pool_token_amount: u64,
```
[the remark this answers is not in the document]
Comment (emmazzz) [reply crates/sui-framework/sources/governance/validator_set.move:200]: > Is this adding the ability to partially switch delegation?

That's right.

> When switch_pool_token_amount is equal to the balance, would we destroy the delegation at epoch boundary?

No. Delegation objects are owned by the delegators and not tracked by us so we don't have access to delegation objects at epoch boundaries. When `switch_pool_token_amount` is equal to the balance, the delegation object will become empty after this transaction. The delegator can then call an entry fun `destroy_empty_delegation` in staking_pool to destroy the empty delegation.

**Comment by emmazzz** [comment]
> Please add more doc comments describing what each function does semantically. I found it a bit difficult to follow.

Will do. Agreed that some functions are getting a bit heavier and more error-prone.

**Comment by emmazzz** [comment]
I added more comments and cleaned up the code a bit in the newer commit. Hope it's easier to follow now. Also for `staking_pool` module, it might be easier to read the code directly instead of looking at the diff. @lxfind

**Review by lxfind — APPROVED** [review]

**Pull request merged**